File: String Benchmark/string-benchmark.rs
use std::io::{BufRead, BufReader};
use std::time::{SystemTime, UNIX_EPOCH};


const TARGET_SIZE: usize = 1024 * 1024;
const WORD_COUNT: usize = 16;
const LINE_LIMIT: usize = 80;
const FNV_OFFSET: u64 = 14695981039346656037;
const FNV_PRIME: u64 = 1099511628211;
const REFERENCE_HASH: u64 = 0x35ce3126ab961070;


fn random_next ( state: &mut u32 ) -> u32
{
	*state = state.wrapping_mul(1664525).wrapping_add(1013904223);
	return *state;
}


fn hash_bytes ( text: &[u8] ) -> u64
{
	let mut hash = FNV_OFFSET;

	for byte in text {
		hash ^= *byte as u64;
		hash = hash.wrapping_mul(FNV_PRIME);
	}
	return hash;
}


fn build_source_text ( words: &[&str] ) -> (String, usize, u64)
{
	let mut text = String::with_capacity(TARGET_SIZE + 64);
	let mut word_count = 0;
	let mut state = 0x12345678u32;

	loop {
		let index = random_next(&mut state) as usize % WORD_COUNT;
		let word = words[index];

		if word_count > 0 {
			text.push(' ');
		}

		text.push_str(word);
		word_count += 1;

		if text.len() > TARGET_SIZE {
			break;
		}
	}

	let hash = hash_bytes(text.as_bytes());
	return (text, word_count, hash);
}


fn wrap_words ( words: &Vec<&str> ) -> String
{
	let mut text = String::with_capacity(TARGET_SIZE + 64);
	let mut line_length = 0;

	for i in 0..words.len() {
		let word = words[i];
		let word_length = word.len();

		if i == 0 {
			text.push_str(word);
			line_length = word_length;
			continue;
		}

		if line_length + 1 + word_length >= LINE_LIMIT {
			text.push('\n');
			text.push_str(word);
			line_length = word_length;
		} else {
			text.push(' ');
			text.push_str(word);
			line_length += 1 + word_length;
		}
	}

	return text;
}


fn hash_wrapped_text ( text: &str ) -> u64
{
	let mut hash = FNV_OFFSET;
	let reader = BufReader::new(text.as_bytes());
	let mut first = true;

	for line in reader.lines() {
		if !first {
			hash ^= b' ' as u64;
			hash = hash.wrapping_mul(FNV_PRIME);
		}
		first = false;

		for byte in line.unwrap().as_bytes() {
			hash ^= *byte as u64;
			hash = hash.wrapping_mul(FNV_PRIME);
		}
	}

	return hash;
}


fn run_benchmark ( source_text: &String, source_hash: u64, runs: i32 )
{
	for _ in 0..runs {
		let split_buffer = source_text.clone();
		let split_words: Vec<&str> = split_buffer.split(' ').collect();
		let wrapped_text = wrap_words(&split_words);
		let wrapped_hash = hash_wrapped_text(&wrapped_text);

		if wrapped_hash != source_hash {
			eprintln!(
				"Hash mismatch: 0x{:016x} != 0x{:016x}",
				wrapped_hash,
				source_hash
			);
			std::process::exit(1);
		}
	}
}


fn main ( )
{
	let words = [
		"I",
		"we",
		"cat",
		"tree",
		"apple",
		"bridge",
		"lantern",
		"mountain",
		"blueberry",
		"basketball",
		"grandfather",
		"microbiology",
		"determination",
		"responsibility",
		"experimentation",
		"counterclockwise"
	];
	let args: Vec<String> = std::env::args().collect();
	let runs = if args.len() <= 1 {
		1
	} else {
		args[1].parse::<i32>().unwrap_or(1)
	};
	let (source_text, _, source_hash) = build_source_text(&words);

	if source_hash != REFERENCE_HASH {
		eprintln!(
			"Source hash mismatch: 0x{:016x} != 0x{:016x}",
			source_hash,
			REFERENCE_HASH
		);
		std::process::exit(1);
	}

	let start = SystemTime::now();
	run_benchmark(&source_text, source_hash, runs);
	let end = SystemTime::now();
	let elapsed = end.duration_since(UNIX_EPOCH).unwrap().as_secs_f64()
		- start.duration_since(UNIX_EPOCH).unwrap().as_secs_f64();
	eprintln!("Rust Elapsed {:.2}", elapsed);
}

File: String Benchmark/string-benchmark.m
#import <Foundation/Foundation.h>

static const NSUInteger targetSize = 1024 * 1024;
static const NSUInteger wordCount = 16;
static const NSUInteger lineLimit = 80;
static const uint64_t fnvOffset = UINT64_C(14695981039346656037);
static const uint64_t fnvPrime = UINT64_C(1099511628211);
static const uint64_t referenceHash = UINT64_C(0xc96d7fcba133ffd5);


static uint32_t randomNext ( uint32_t * state )
{
	*state = *state * UINT32_C(1664525) + UINT32_C(1013904223);
	return *state;
}


static uint64_t hashUTF16 ( NSString * text )
{
	uint64_t hash = fnvOffset;
	NSUInteger i;

	for (i = 0; i < [text length]; i++) {
		hash ^= [text characterAtIndex:i];
		hash *= fnvPrime;
	}
	return hash;
}


static NSDictionary * buildBenchmarkData ( void )
{
	NSArray * words = @[
		@"I",
		@"we",
		@"cat",
		@"tree",
		@"café",
		@"naïve",
		@"jalapeño",
		@"mountain",
		@"blueberry",
		@"basketball",
		@"grandfather",
		@"microbiology",
		@"determination",
		@"responsibility",
		@"experimentation",
		@"counterclockwise"
	];
	NSMutableString * text = [NSMutableString stringWithCapacity:
		targetSize + 64];
	NSUInteger length = 0;
	NSUInteger generatedWordCount = 0;
	uint32_t state = UINT32_C(0x12345678);
	uint64_t sourceHash;

	while (1) {
		NSString * word = words[randomNext(&state) % wordCount];
		NSUInteger wordLength = [word length];

		if (generatedWordCount > 0) {
			[text appendString:@" "];
			length++;
		}

		[text appendString:word];
		length += wordLength;
		generatedWordCount++;

		if (length > targetSize) {
			break;
		}
	}

	sourceHash = hashUTF16(text);
	if (sourceHash != referenceHash) {
		fprintf(
			stderr,
			"Source hash mismatch: 0x%016llx != 0x%016llx\n",
			sourceHash,
			referenceHash
		);
		exit(1);
	}

	return @{
		@"sourceText": text,
		@"sourceHash": @(sourceHash)
	};
}


static NSString * wrapWords ( NSArray * words )
{
	NSMutableString * text = [NSMutableString stringWithCapacity:
		targetSize + 64];
	NSUInteger lineLength = 0;
	NSUInteger i;

	for (i = 0; i < [words count]; i++) {
		NSString * word = words[i];
		NSUInteger wordLength = [word length];

		if (i == 0) {
			[text appendString:word];
			lineLength = wordLength;
			continue;
		}

		if (lineLength + 1 + wordLength >= lineLimit) {
			[text appendString:@"\n"];
			[text appendString:word];
			lineLength = wordLength;
		} else {
			[text appendString:@" "];
			[text appendString:word];
			lineLength += 1 + wordLength;
		}
	}

	return text;
}


static uint64_t hashWrappedText ( NSString * text )
{
	__block uint64_t hash = fnvOffset;
	__block BOOL first = YES;

	[text enumerateLinesUsingBlock:
		^(NSString * line, BOOL * stop __attribute__((unused))) {
			NSUInteger i;

			if (!first) {
				hash ^= ' ';
				hash *= fnvPrime;
			}
			first = NO;

			for (i = 0; i < [line length]; i++) {
				hash ^= [line characterAtIndex:i];
				hash *= fnvPrime;
			}
		}
	];

	return hash;
}


static void runBenchmark ( NSDictionary * data, int runs )
{
	NSString * sourceText = data[@"sourceText"];
	uint64_t sourceHash = [data[@"sourceHash"] unsignedLongLongValue];
	int i;

	for (i = 0; i < runs; i++) {
		@autoreleasepool {
			NSString * splitBuffer = [[NSString alloc] initWithString:
				sourceText];
			NSArray * splitWords = [splitBuffer componentsSeparatedByString:
				@" "];
			NSString * wrappedText = wrapWords(splitWords);
			uint64_t wrappedHash = hashWrappedText(wrappedText);

			if (wrappedHash != sourceHash) {
				fprintf(
					stderr,
					"Hash mismatch: 0x%016llx != 0x%016llx\n",
					wrappedHash,
					sourceHash
				);
				exit(1);
			}
		}
	}
}


int main ( int argc, const char * argv[] )
{
	NSDictionary * data;
	double startTime;
	double timeElapsed;
	int runs = 1;

	if (argc > 1) {
		sscanf(argv[1], "%d", &runs);
	}

	@autoreleasepool {
		data = buildBenchmarkData();
		startTime = CFAbsoluteTimeGetCurrent();
		runBenchmark(data, runs);
		timeElapsed = CFAbsoluteTimeGetCurrent() - startTime;
		fprintf(stderr, "Objective-C Elapsed %.3f\n", timeElapsed);
	}

	return 0;
}
